File: Duct/Core/ChildReconciler.cs
using Microsoft.UI.Xaml;
using WinUI = Microsoft.UI.Xaml.Controls;

namespace Duct.Core;

/// <summary>
/// Keyed child reconciliation using Longest Increasing Subsequence (LIS).
///
/// Strategies:
///   1. Unkeyed children: positional reconciliation (match by index)
///   2. Keyed children: prefix/suffix stripping + LIS for minimal moves
/// </summary>
internal static class ChildReconciler
{
    /// <summary>
    /// Reconciles old and new child element arrays against a Panel's Children collection.
    /// </summary>
    internal static void Reconcile(
        Element[] oldChildren,
        Element[] newChildren,
        IChildCollection children,
        Reconciler reconciler,
        Action requestRerender)
    {
        // Filter out nulls and EmptyElements
        var oldFiltered = Filter(oldChildren);
        var newFiltered = Filter(newChildren);

        bool hasKeys = HasAnyKeys(oldFiltered) || HasAnyKeys(newFiltered);

        if (hasKeys)
            ReconcileKeyed(oldFiltered, newFiltered, children, reconciler, requestRerender);
        else
            ReconcilePositional(oldFiltered, newFiltered, children, reconciler, requestRerender);
    }

    /// <summary>
    /// Positional reconciliation: match children by index.
    /// O(max(old, new)) — no reorder detection, but simple and fast.
    /// </summary>
    private static void ReconcilePositional(
        Element[] oldChildren,
        Element[] newChildren,
        IChildCollection children,
        Reconciler reconciler,
        Action requestRerender)
    {
        int common = Math.Min(oldChildren.Length, newChildren.Length);

        // Update common children in place
        for (int i = 0; i < common; i++)
        {
            if (i >= children.Count) break;

            if (reconciler.CanUpdate(oldChildren[i], newChildren[i]))
            {
                var replacement = reconciler.UpdateChild(oldChildren[i], newChildren[i], children.Get(i), requestRerender);
                if (replacement is not null)
                {
                    // Child type changed at runtime — replace in place
                    reconciler.UnmountChild(children.Get(i));
                    children.Replace(i, replacement);
                }
            }
            else
            {
                // Type mismatch — unmount old, mount new
                reconciler.UnmountChild(children.Get(i));
                var newControl = reconciler.Mount(newChildren[i], requestRerender);
                if (newControl is not null)
                    children.Replace(i, newControl);
            }
        }

        // Remove excess old children (from end to start to keep indices stable)
        for (int i = children.Count - 1; i >= common; i--)
        {
            var old = children.Get(i);
            children.RemoveAt(i);
            reconciler.UnmountAndPool(old);
        }

        // Insert new children beyond old count
        for (int i = common; i < newChildren.Length; i++)
        {
            var ctrl = reconciler.Mount(newChildren[i], requestRerender);
            if (ctrl is not null)
                children.Insert(children.Count, ctrl);
        }
    }

    /// <summary>
    /// Keyed reconciliation using prefix/suffix stripping + LIS.
    /// Minimizes DOM operations for reordered lists.
    /// </summary>
    private static void ReconcileKeyed(
        Element[] oldChildren,
        Element[] newChildren,
        IChildCollection children,
        Reconciler reconciler,
        Action requestRerender)
    {
        int oldLen = oldChildren.Length;
        int newLen = newChildren.Length;

        // Phase 1: Common prefix
        int prefixLen = 0;
        while (prefixLen < oldLen && prefixLen < newLen &&
               KeyMatch(oldChildren[prefixLen], newChildren[prefixLen]) &&
               reconciler.CanUpdate(oldChildren[prefixLen], newChildren[prefixLen]))
        {
            // Update in place
            if (prefixLen < children.Count)
            {
                var replacement = reconciler.UpdateChild(oldChildren[prefixLen], newChildren[prefixLen], children.Get(prefixLen), requestRerender);
                if (replacement is not null)
                {
                    reconciler.UnmountChild(children.Get(prefixLen));
                    children.Replace(prefixLen, replacement);
                }
            }
            prefixLen++;
        }

        // Phase 2: Common suffix
        int suffixLen = 0;
        while (suffixLen < (oldLen - prefixLen) && suffixLen < (newLen - prefixLen) &&
               KeyMatch(oldChildren[oldLen - 1 - suffixLen], newChildren[newLen - 1 - suffixLen]) &&
               reconciler.CanUpdate(oldChildren[oldLen - 1 - suffixLen], newChildren[newLen - 1 - suffixLen]))
        {
            // Update in place (from end)
            int oldIdx = oldLen - 1 - suffixLen;
            int panelIdx = children.Count - 1 - suffixLen;
            if (panelIdx >= 0 && panelIdx < children.Count)
            {
                var replacement = reconciler.UpdateChild(oldChildren[oldIdx], newChildren[newLen - 1 - suffixLen], children.Get(panelIdx), requestRerender);
                if (replacement is not null)
                {
                    reconciler.UnmountChild(children.Get(panelIdx));
                    children.Replace(panelIdx, replacement);
                }
            }
            suffixLen++;
        }

        // Phase 3: Middle section
        int oldStart = prefixLen;
        int oldEnd = oldLen - suffixLen;
        int newStart = prefixLen;
        int newEnd = newLen - suffixLen;

        int oldMidLen = oldEnd - oldStart;
        int newMidLen = newEnd - newStart;

        if (oldMidLen == 0 && newMidLen == 0)
            return; // Prefix + suffix covered everything

        if (oldMidLen == 0)
        {
            // Only insertions
            for (int i = 0; i < newMidLen; i++)
            {
                var ctrl = reconciler.Mount(newChildren[newStart + i], requestRerender);
                if (ctrl is not null)
                    children.Insert(prefixLen + i, ctrl);
            }
            return;
        }

        if (newMidLen == 0)
        {
            // Only removals (from end to start)
            for (int i = oldMidLen - 1; i >= 0; i--)
            {
                int panelIdx = prefixLen + i;
                if (panelIdx < children.Count)
                {
                    var old = children.Get(panelIdx);
                    children.RemoveAt(panelIdx);
                    reconciler.UnmountAndPool(old);
                }
            }
            return;
        }

        // Middle section requires key mapping + LIS
        ReconcileKeyedMiddle(oldChildren, newChildren, oldStart, oldMidLen, newStart, newMidLen,
            prefixLen, suffixLen, children, reconciler, requestRerender);
    }

    /// <summary>
    /// Keyed middle section reconciliation using LIS for minimal moves.
    /// </summary>
    private static void ReconcileKeyedMiddle(
        Element[] oldChildren, Element[] newChildren,
        int oldStart, int oldMidLen, int newStart, int newMidLen,
        int prefixLen, int suffixLen,
        IChildCollection children,
        Reconciler reconciler,
        Action requestRerender)
    {
        // Build old key → index map
        var oldKeyMap = new Dictionary<string, int>(oldMidLen);
        for (int i = 0; i < oldMidLen; i++)
        {
            var key = GetKey(oldChildren[oldStart + i], oldStart + i);
            oldKeyMap[key] = i;
        }

        // Map new keys to old indices
        var newToOld = new int[newMidLen];
        var matched = new bool[oldMidLen];
        for (int i = 0; i < newMidLen; i++)
        {
            var key = GetKey(newChildren[newStart + i], newStart + i);
            if (oldKeyMap.TryGetValue(key, out int oldIdx) &&
                reconciler.CanUpdate(oldChildren[oldStart + oldIdx], newChildren[newStart + i]))
            {
                newToOld[i] = oldIdx;
                matched[oldIdx] = true;
            }
            else
            {
                newToOld[i] = -1;
            }
        }

        // Compute LIS on newToOld
        var lisIndices = ComputeLIS(newToOld);
        var inLIS = new HashSet<int>(lisIndices);

        // Step 1: Remove unmatched old items (reverse order for stable indices)
        for (int i = oldMidLen - 1; i >= 0; i--)
        {
            if (!matched[i])
            {
                int panelIdx = prefixLen + i;
                if (panelIdx < children.Count)
                {
                    var old = children.Get(panelIdx);
                    children.RemoveAt(panelIdx);
                    reconciler.UnmountAndPool(old);
                }
            }
        }

        // Step 2: Process new items - insert new, move existing not in LIS
        for (int i = 0; i < newMidLen; i++)
        {
            int targetPanelIdx = prefixLen + i;

            if (newToOld[i] == -1)
            {
                var ctrl = reconciler.Mount(newChildren[newStart + i], requestRerender);
                if (ctrl is not null)
                    children.Insert(targetPanelIdx, ctrl);
            }
            else if (inLIS.Contains(i))
            {
                if (targetPanelIdx < children.Count)
                {
                    var replacement = reconciler.UpdateChild(
                        oldChildren[oldStart + newToOld[i]],
                        newChildren[newStart + i],
                        children.Get(targetPanelIdx),
                        requestRerender);
                    if (replacement is not null)
                    {
                        reconciler.UnmountChild(children.Get(targetPanelIdx));
                        children.Replace(targetPanelIdx, replacement);
                    }
                }
            }
            else
            {
                int oldRelIdx = newToOld[i];
                int currentPos = FindItemByOldIndex(children, oldChildren, oldStart + oldRelIdx, prefixLen, children.Count - suffixLen, reconciler);
                if (currentPos >= 0 && currentPos != targetPanelIdx)
                {
                    children.Move(currentPos, targetPanelIdx);
                }
                if (targetPanelIdx < children.Count)
                {
                    var replacement = reconciler.UpdateChild(
                        oldChildren[oldStart + oldRelIdx],
                        newChildren[newStart + i],
                        children.Get(targetPanelIdx),
                        requestRerender);
                    if (replacement is not null)
                    {
                        reconciler.UnmountChild(children.Get(targetPanelIdx));
                        children.Replace(targetPanelIdx, replacement);
                    }
                }
            }
        }
    }

    /// <summary>
    /// Find the current position of an item that was originally at a given old index.
    /// Uses the reconciler's element map to match elements.
    /// </summary>
    private static int FindItemByOldIndex(
        IChildCollection children,
        Element[] oldElements,
        int oldIndex,
        int searchStart,
        int searchEnd,
        Reconciler reconciler)
    {
        for (int i = searchStart; i < searchEnd && i < children.Count; i++)
        {
            var child = children.Get(i);
            if (child is FrameworkElement fe && fe.Tag is Element tagElement)
            {
                if (oldIndex < oldElements.Length &&
                    GetKey(tagElement, -1) == GetKey(oldElements[oldIndex], oldIndex))
                    return i;
            }
        }
        return -1;
    }

    /// <summary>
    /// Compute Longest Increasing Subsequence.
    /// Returns indices into the input array that form the LIS.
    /// Skips entries with value -1 (unmapped items).
    /// O(n log n) using patience sorting.
    /// </summary>
    internal static HashSet<int> ComputeLIS(int[] arr)
    {
        if (arr.Length == 0) return new HashSet<int>();

        var tails = new List<int>();     // Smallest tail values
        var tailIndices = new List<int>(); // Indices in arr corresponding to tails
        var predecessors = new int[arr.Length];
        Array.Fill(predecessors, -1);

        for (int i = 0; i < arr.Length; i++)
        {
            if (arr[i] == -1) continue; // Skip unmapped

            int val = arr[i];

            // Binary search for insertion position
            int lo = 0, hi = tails.Count;
            while (lo < hi)
            {
                int mid = (lo + hi) / 2;
                if (tails[mid] < val) lo = mid + 1;
                else hi = mid;
            }

            if (lo == tails.Count)
            {
                tails.Add(val);
                tailIndices.Add(i);
            }
            else
            {
                tails[lo] = val;
                tailIndices[lo] = i;
            }

            if (lo > 0)
                predecessors[i] = tailIndices[lo - 1];
        }

        // Backtrack to find actual LIS indices
        var result = new HashSet<int>();
        if (tailIndices.Count == 0) return result;

        int idx = tailIndices[^1];
        while (idx != -1)
        {
            result.Add(idx);
            idx = predecessors[idx];
        }

        return result;
    }

    private static Element[] Filter(Element[] elements)
    {
        // Fast path: if no filtering needed, return as-is
        bool needsFilter = false;
        for (int i = 0; i < elements.Length; i++)
        {
            if (elements[i] is null or EmptyElement) { needsFilter = true; break; }
        }
        if (!needsFilter) return elements;

        var result = new List<Element>(elements.Length);
        for (int i = 0; i < elements.Length; i++)
        {
            if (elements[i] is not null and not EmptyElement)
                result.Add(elements[i]);
        }
        return result.ToArray();
    }

    private static bool HasAnyKeys(Element[] elements)
    {
        for (int i = 0; i < elements.Length; i++)
        {
            if (elements[i].Key is not null) return true;
        }
        return false;
    }

    private static bool KeyMatch(Element a, Element b)
    {
        // Both must have the same key (or both null) AND same type
        if (a.GetType() != b.GetType()) return false;
        return a.Key == b.Key;
    }

    private static string GetKey(Element element, int positionalIndex)
    {
        // Use explicit key if available, otherwise fall back to type+position
        if (element.Key is not null) return element.Key;
        return $"__pos_{positionalIndex}_{element.GetType().Name}";
    }
}

File: DuctCpp/DuctCpp/src/child_reconciler.cpp
#include "pch.h"
#include "child_reconciler.h"
#include "reconciler.h"
#include <unordered_map>
#include <algorithm>
#include <sstream>

namespace duct {

// --- Helpers ---

std::vector<const Element*> ChildReconciler::filter(const std::vector<Element>& elements) {
    std::vector<const Element*> result;
    result.reserve(elements.size());
    for (const auto& el : elements) {
        if (!std::holds_alternative<EmptyElement>(el.data)) {
            result.push_back(&el);
        }
    }
    return result;
}

bool ChildReconciler::has_any_keys(const std::vector<const Element*>& elements) {
    for (const auto* el : elements) {
        if (el->key.has_value()) return true;
    }
    return false;
}

bool ChildReconciler::key_match(const Element& a, const Element& b) {
    // Both must have the same type AND same key (or both no key)
    if (a.data.index() != b.data.index()) return false;
    return a.key == b.key;
}

std::string ChildReconciler::get_key(const Element& element, int positional_index) {
    if (element.key.has_value()) return *element.key;
    return "__pos_" + std::to_string(positional_index) + "_" + std::to_string(element.data.index());
}

// --- Main Entry Point ---

void ChildReconciler::reconcile(
    const std::vector<Element>& old_children,
    const std::vector<Element>& new_children,
    IChildCollection& children,
    Reconciler& reconciler,
    std::function<void()> request_rerender)
{
    auto old_filtered = filter(old_children);
    auto new_filtered = filter(new_children);

    bool has_keys = has_any_keys(old_filtered) || has_any_keys(new_filtered);

    if (has_keys)
        reconcile_keyed(old_filtered, new_filtered, children, reconciler, request_rerender);
    else
        reconcile_positional(old_filtered, new_filtered, children, reconciler, request_rerender);
}

// --- Positional Reconciliation ---

void ChildReconciler::reconcile_positional(
    const std::vector<const Element*>& old_children,
    const std::vector<const Element*>& new_children,
    IChildCollection& children,
    Reconciler& reconciler,
    std::function<void()> request_rerender)
{
    int old_count = static_cast<int>(old_children.size());
    int new_count = static_cast<int>(new_children.size());
    int common = std::min(old_count, new_count);

    // Update common children in place
    for (int i = 0; i < common; i++) {
        if (i >= children.count()) break;

        if (Reconciler::can_update(*old_children[i], *new_children[i])) {
            auto replacement = reconciler.update(*old_children[i], *new_children[i],
                                                  children.get(i), request_rerender);
            if (replacement) {
                // update() returned a replacement control
                reconciler.unmount(children.get(i));
                children.replace(i, replacement);
            }
        } else {
            // Type mismatch — unmount old, mount new
            reconciler.unmount(children.get(i));
            auto new_control = reconciler.mount(*new_children[i], request_rerender);
            if (new_control)
                children.replace(i, new_control);
        }
    }

    // Remove excess old children (from end to start for stable indices)
    for (int i = children.count() - 1; i >= common; i--) {
        auto old_ctrl = children.get(i);
        children.remove_at(i);
        reconciler.unmount(old_ctrl);
    }

    // Insert new children beyond old count
    for (int i = common; i < new_count; i++) {
        auto ctrl = reconciler.mount(*new_children[i], request_rerender);
        if (ctrl)
            children.insert(children.count(), ctrl);
    }
}

// --- Keyed Reconciliation ---

void ChildReconciler::reconcile_keyed(
    const std::vector<const Element*>& old_children,
    const std::vector<const Element*>& new_children,
    IChildCollection& children,
    Reconciler& reconciler,
    std::function<void()> request_rerender)
{
    int old_len = static_cast<int>(old_children.size());
    int new_len = static_cast<int>(new_children.size());

    // Phase 1: Common prefix
    int prefix_len = 0;
    while (prefix_len < old_len && prefix_len < new_len &&
           key_match(*old_children[prefix_len], *new_children[prefix_len]) &&
           Reconciler::can_update(*old_children[prefix_len], *new_children[prefix_len]))
    {
        if (prefix_len < children.count()) {
            auto replacement = reconciler.update(*old_children[prefix_len], *new_children[prefix_len],
                                                  children.get(prefix_len), request_rerender);
            if (replacement) {
                reconciler.unmount(children.get(prefix_len));
                children.replace(prefix_len, replacement);
            }
        }
        prefix_len++;
    }

    // Phase 2: Common suffix
    int suffix_len = 0;
    while (suffix_len < (old_len - prefix_len) && suffix_len < (new_len - prefix_len) &&
           key_match(*old_children[old_len - 1 - suffix_len], *new_children[new_len - 1 - suffix_len]) &&
           Reconciler::can_update(*old_children[old_len - 1 - suffix_len], *new_children[new_len - 1 - suffix_len]))
    {
        int old_idx = old_len - 1 - suffix_len;
        int panel_idx = children.count() - 1 - suffix_len;
        if (panel_idx >= 0 && panel_idx < children.count()) {
            auto replacement = reconciler.update(*old_children[old_idx], *new_children[new_len - 1 - suffix_len],
                                                  children.get(panel_idx), request_rerender);
            if (replacement) {
                reconciler.unmount(children.get(panel_idx));
                children.replace(panel_idx, replacement);
            }
        }
        suffix_len++;
    }

    // Phase 3: Middle section
    int old_start = prefix_len;
    int old_end = old_len - suffix_len;
    int new_start = prefix_len;
    int new_end = new_len - suffix_len;

    int old_mid_len = old_end - old_start;
    int new_mid_len = new_end - new_start;

    if (old_mid_len == 0 && new_mid_len == 0)
        return; // Prefix + suffix covered everything

    if (old_mid_len == 0) {
        // Only insertions
        for (int i = 0; i < new_mid_len; i++) {
            auto ctrl = reconciler.mount(*new_children[new_start + i], request_rerender);
            if (ctrl)
                children.insert(prefix_len + i, ctrl);
        }
        return;
    }

    if (new_mid_len == 0) {
        // Only removals (from end to start)
        for (int i = old_mid_len - 1; i >= 0; i--) {
            int panel_idx = prefix_len + i;
            if (panel_idx < children.count()) {
                auto old_ctrl = children.get(panel_idx);
                children.remove_at(panel_idx);
                reconciler.unmount(old_ctrl);
            }
        }
        return;
    }

    // Middle section requires key mapping + LIS
    reconcile_keyed_middle(old_children, new_children, old_start, old_mid_len, new_start, new_mid_len,
                           prefix_len, suffix_len, children, reconciler, request_rerender);
}

// --- Keyed Middle Section (LIS) ---

void ChildReconciler::reconcile_keyed_middle(
    const std::vector<const Element*>& old_children,
    const std::vector<const Element*>& new_children,
    int old_start, int old_mid_len,
    int new_start, int new_mid_len,
    int prefix_len, int suffix_len,
    IChildCollection& children,
    Reconciler& reconciler,
    std::function<void()> request_rerender)
{
    // Build old key → index map
    std::unordered_map<std::string, int> old_key_map;
    old_key_map.reserve(old_mid_len);
    for (int i = 0; i < old_mid_len; i++) {
        auto key = get_key(*old_children[old_start + i], old_start + i);
        old_key_map[key] = i;
    }

    // Map new keys to old indices
    std::vector<int> new_to_old(new_mid_len);
    std::vector<bool> matched(old_mid_len, false);
    for (int i = 0; i < new_mid_len; i++) {
        auto key = get_key(*new_children[new_start + i], new_start + i);
        auto it = old_key_map.find(key);
        if (it != old_key_map.end() &&
            Reconciler::can_update(*old_children[old_start + it->second], *new_children[new_start + i]))
        {
            new_to_old[i] = it->second;
            matched[it->second] = true;
        } else {
            new_to_old[i] = -1;
        }
    }

    // Compute LIS on new_to_old
    auto lis_indices = compute_lis(new_to_old);

    // Step 1: Remove unmatched old items (reverse order for stable indices)
    for (int i = old_mid_len - 1; i >= 0; i--) {
        if (!matched[i]) {
            int panel_idx = prefix_len + i;
            if (panel_idx < children.count()) {
                auto old_ctrl = children.get(panel_idx);
                children.remove_at(panel_idx);
                reconciler.unmount(old_ctrl);
            }
        }
    }

    // Step 2: Process new items - insert new, move existing not in LIS
    for (int i = 0; i < new_mid_len; i++) {
        int target_panel_idx = prefix_len + i;

        if (new_to_old[i] == -1) {
            // New item — mount and insert
            auto ctrl = reconciler.mount(*new_children[new_start + i], request_rerender);
            if (ctrl)
                children.insert(target_panel_idx, ctrl);
        }
        else if (lis_indices.count(i)) {
            // In LIS — update in place (no move needed)
            if (target_panel_idx < children.count()) {
                auto replacement = reconciler.update(
                    *old_children[old_start + new_to_old[i]],
                    *new_children[new_start + i],
                    children.get(target_panel_idx),
                    request_rerender);
                if (replacement) {
                    reconciler.unmount(children.get(target_panel_idx));
                    children.replace(target_panel_idx, replacement);
                }
            }
        }
        else {
            // Not in LIS — needs to be moved
            int old_rel_idx = new_to_old[i];
            int current_pos = find_item_by_old_index(
                children, old_children, old_start + old_rel_idx,
                prefix_len, children.count() - suffix_len, reconciler);
            if (current_pos >= 0 && current_pos != target_panel_idx) {
                children.move(current_pos, target_panel_idx);
            }
            if (target_panel_idx < children.count()) {
                auto replacement = reconciler.update(
                    *old_children[old_start + old_rel_idx],
                    *new_children[new_start + i],
                    children.get(target_panel_idx),
                    request_rerender);
                if (replacement) {
                    reconciler.unmount(children.get(target_panel_idx));
                    children.replace(target_panel_idx, replacement);
                }
            }
        }
    }
}

// --- Find Item by Old Index ---

int ChildReconciler::find_item_by_old_index(
    IChildCollection& children,
    const std::vector<const Element*>& old_elements,
    int old_index,
    int search_start, int search_end,
    Reconciler& /*reconciler*/)
{
    for (int i = search_start; i < search_end && i < children.count(); i++) {
        auto child = children.get(i);
        if (auto fe = child.try_as<xaml::FrameworkElement>()) {
            auto tag = fe.Tag();
            if (tag) {
                // Tag stores a pointer to the Element as a uint64
                auto ptr_val = winrt::unbox_value_or<uint64_t>(tag, 0);
                if (ptr_val != 0) {
                    auto* tag_el = reinterpret_cast<const Element*>(ptr_val);
                    if (old_index < static_cast<int>(old_elements.size()) &&
                        get_key(*tag_el, -1) == get_key(*old_elements[old_index], old_index))
                        return i;
                }
            }
        }
    }
    return -1;
}

// --- Longest Increasing Subsequence ---

std::unordered_set<int> ChildReconciler::compute_lis(const std::vector<int>& arr) {
    if (arr.empty()) return {};

    std::vector<int> tails;        // Smallest tail values
    std::vector<int> tail_indices; // Indices in arr corresponding to tails
    std::vector<int> predecessors(arr.size(), -1);

    for (int i = 0; i < static_cast<int>(arr.size()); i++) {
        if (arr[i] == -1) continue; // Skip unmapped

        int val = arr[i];

        // Binary search for insertion position
        int lo = 0, hi = static_cast<int>(tails.size());
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (tails[mid] < val) lo = mid + 1;
            else hi = mid;
        }

        if (lo == static_cast<int>(tails.size())) {
            tails.push_back(val);
            tail_indices.push_back(i);
        } else {
            tails[lo] = val;
            tail_indices[lo] = i;
        }

        if (lo > 0)
            predecessors[i] = tail_indices[lo - 1];
    }

    // Backtrack to find actual LIS indices
    std::unordered_set<int> result;
    if (tail_indices.empty()) return result;

    int idx = tail_indices.back();
    while (idx != -1) {
        result.insert(idx);
        idx = predecessors[idx];
    }

    return result;
}

} // namespace duct
